### PR TITLE
#50  Adding career path

### PR DESCRIPTION
**Pull request by Werfit**

### CODE REVIEW

File: backend/src/data/migrations/1644929911066-skill-fields-update.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    await queryRunner.query(
+      'CREATE TYPE "public"."skill_type_enum" AS ENUM(\'Soft skills\', \'Hard skills\', \'Language\')',
+    );
+    await queryRunner.query(
+      'ALTER TABLE "skill" ADD "type" "public"."skill_type_enum" NOT NULL',
+    );
+    await queryRunner.query(
+      'ALTER TABLE "skill" ADD "name" character varying(250) NOT NULL',
+    );
+    // await queryRunner.query(`ALTER TABLE "skill" DROP COLUMN "name"`);
```
Comment (aleksandr-kuharenko) [inline backend/src/data/migrations/1644929911066-skill-fields-update.ts]: ?

File: backend/src/api/controllers/career-path.controller.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+export const createCareerPath = async (
+  data: CareerPath,
+  company: Company,
+): Promise<CareeerPathType> => {
+  const domain = await createDomain(data.domain, company);
+  const domainLevels = Object.entries(data.level);
+
+  const domainLevelInstances: DomainLevel[] = [];
+
+  await asyncForEach(async ([name, value]) => {
```
Comment (karyna-kh) [inline backend/src/api/controllers/career-path.controller.ts]: remove nesting and move each asyncForEach() function call to separate functions

File: backend/src/data/seeders/skills.seeder.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
+import { getCustomRepository } from 'typeorm';
+
+import { skills } from '../seed-data/skills.data';
+import { companies } from '../seed-data/company.data';
+
+import { Skill } from '../entities/skill';
+import { asyncForEach } from '../../common/helpers/array.helper';
+
+import CompanyRepository from '../repositories/company.repository';
+// import { PermissionType } from 'growup-shared';
```
Comment (aleksandr-kuharenko) [inline backend/src/data/seeders/skills.seeder.ts]:
```suggestion
```

File: backend/src/api/controllers/career-path.controller.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    const level = await createDomainLevel({
+      name,
+      domain,
+      prev: domainLevelInstances.length ? domainLevelInstances.at(-1) : null,
+    });
+
+    const categories = Object.entries(value);
+
+    asyncForEach(async ([name, value]) => {
```
Comment (aleksandr-kuharenko) [inline backend/src/api/controllers/career-path.controller.ts]:
```suggestion
    await asyncForEach(async ([name, value]) => {
```

File: backend/src/api/controllers/career-path.controller.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+      const skills = value;
+
+      asyncForEach(async (skill) => {
```
Comment (aleksandr-kuharenko) [inline backend/src/api/controllers/career-path.controller.ts]:
```suggestion
      await asyncForEach(async (skill) => {
```

File: backend/src/data/migrations/1644920948847-add-column-name-domain-level.ts [hunk 6]
```diff
@@ -0,0 +1,17 @@
+import { MigrationInterface, QueryRunner } from 'typeorm';
+
+export class addColumnNameDomainLevel1644920948847
```
Comment (karyna-kh) [inline backend/src/data/migrations/1644920948847-add-column-name-domain-level.ts:3]: are such function names allowed?

File: backend/src/data/migrations/1644929911066-skill-fields-update.ts (hunk 1 above)
Comment (karyna-kh) [inline backend/src/data/migrations/1644929911066-skill-fields-update.ts]: remove commented code

File: backend/src/data/seed-data/skills.data.ts [hunk 7]
```diff
@@ -0,0 +1,20 @@
+import { SkillType } from '~/common/enums/skill-type';
+
+export const skills = [
+  {
+    name: 'HTML5',
```
Comment (karyna-kh) [inline backend/src/data/seed-data/skills.data.ts:5]: move skill names to enum

File: backend/src/data/seeders/skills.seeder.ts (hunk 3 above)
Comment (karyna-kh) [inline backend/src/data/seeders/skills.seeder.ts]: remove commented code

**Review by aleksandr-kuharenko — CHANGES_REQUESTED** [review]

File: backend/src/api/routes/career-path.ts [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+  createCareerPath,
+  getDomainAndLevels,
+} from '../controllers/career-path.controller';
+
+const router = Router();
+
+router.post(
+  '/',
+  run(async (req: Request) => {
+    const company = await getCompany();
```
Comment (aleksandr-kuharenko) [inline backend/src/api/routes/career-path.ts]: find company with id from token. req.companyId

File: backend/src/api/routes/index.ts [hunk 9]
```diff
@@ -4,12 +4,16 @@ import { errorHandlerMiddleware } from '../middlewares/error-handler-middleware'
...
+import careerPathRoute from './career-path';
 import adminRoute from './admin-route';
 import workQuizRoute from './work-quiz';
 
 const routes = (app: Express): void => {
   app.use('/api/auth', authenticationRoute);
   app.use('/api/user', userRoute);
+  app.use('/api/career-path', careerPathRoute);
+
+  app.use('/company', companyRoute);
```
Comment (aleksandr-kuharenko) [inline backend/src/api/routes/index.ts]:
```suggestion
```

**Review by nikita-remeslov — APPROVED** [review]

**Review by aleksandr-kuharenko — CHANGES_REQUESTED** [review]

**Review by aleksandr-kuharenko — APPROVED** [review]

**Pull request merged**